streams/src/main/java/org/apache/kafka/streams/kstream/internals/NamedInternal.java

```diff
@@ -0,0 +1,127 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.kstream.internals;
+
+import org.apache.kafka.streams.errors.InvalidProcessorException;
+import org.apache.kafka.streams.kstream.Named;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+public class NamedInternal implements Named {
+
+    private static final NamedInternal EMPTY = new NamedInternal();
+
+    private static final int MAX_NAME_LENGTH = 249;
+
+    private final Optional<String> internal;
+
+    public static NamedInternal empty() {
+        return EMPTY;
+    }
+
+    public static NamedInternal with(final String name) {
+        return new NamedInternal(name);
+    }
+
+    /**
+     * Creates a new {@link NamedInternal} instance.
+     * @param internal  the internal name.
+     */
+    public NamedInternal(final Named internal) {
+        this(Objects.requireNonNull(internal).name());
+    }
+
+    public NamedInternal(final String internal) {
+        this.internal = Optional.ofNullable(internal);
+    }
+
+    private NamedInternal() {
+        this.internal = Optional.empty();
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * @throws IllegalStateException if internal name is null.
+     * @return  a string name.
+     */
+    @Override
+    public String name() {
+        if (internal.isPresent()) {
+            return internal.get();
+        } else {
+            throw new IllegalStateException("name is null");
+        }
+    }
+
+    public String suffixWithOrElseGet(final String suffix, final Supplier<String> supplier) {
+        return internal.map(s -> s + suffix).orElseGet(supplier);
+    }
+
+    public String orElseGenerateWithPrefix(final InternalNameProvider provider, final String prefix) {
+        return orElseGet(() -> provider.newProcessorName(prefix));
+    }
+
+    /**
+     * Returns the internal name or the value returns from the supplier.
+     *
+     * @param supplier  the supplier to be used if internal name is empty.
+     * @return an internal string name.
+     */
+    private String orElseGet(final Supplier<String> supplier) {
+        return internal.orElseGet(supplier);
+    }
+
+    /**
+     * Validate the specified processor name. Because the processor name can be used to derive a state store name,
+     * the name is validate against the same rules than for a topic name.
+     * @param processorName the processor name to validate.
+     *
+     * @throws InvalidProcessorException if the processor name is invalid.
+     */
+    public static void validate(final String processorName) {
```

Comment:
Making it a public function would require us to call `validate` on all the places scattered across KTableImpl / KStreamImpl etc classes. Instead I'd suggest we just make it a private function and call it in

1) constructor which takes the String or Named.
2) static function that creates a Named from String.

By doing this we do not need to call this elsewhere, this is because for library created names internally, we are sure that they are always valid so we do not need to validate them.

Reply:
BTW: In that case, we would allow the internal name to be null.